File: src/functional/haskell_core/omni_uot_entropy.hs
-- Omni UoT Pure Entropy (Haskell)
-- Ref: zhiyuanhubj/UoT — NeurIPS 2024
module Omni.UoT.Entropy where
entropy :: [Double] -> Double
entropy probs = negate $ sum [p * logBase 2 p | p <- probs, p > 1e-12]
informationGain :: [Double] -> [Double] -> Double
informationGain prior posterior = max 0 (entropy prior - entropy posterior)
selectBest :: [[Double]] -> [Double] -> (Int, Double)
selectBest relevances hyps = foldl (\(bi,bg) (i,r) ->
  let post = zipWith (*) hyps r
      s = sum post
      norm = map (/ max s 1e-9) post
      ig = informationGain hyps norm
  in if ig > bg then (i, ig) else (bi, bg)) (0, -1) (zip [0..] relevances)

File: src/system/zig_core/omni_uot_entropy.zig
// Omni UoT Entropy Kernel (Zig)
// Ref: zhiyuanhubj/UoT — NeurIPS 2024
const std = @import("std");
const math = std.math;
pub fn entropy(probs: []const f64) f64 {
    var h: f64 = 0;
    for (probs) |p| {
        if (p > 1e-12) h -= p * math.log2(p);
    }
    return h;
}
pub fn information_gain(prior: []const f64, posterior: []const f64) f64 {
    const h_prior = entropy(prior);
    const h_post = entropy(posterior);
    return if (h_prior > h_post) h_prior - h_post else 0;
}
